addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/WebCacheDeceptionScanRule.java

```diff
@@ -0,0 +1,183 @@
+                            .setMethod(authorisedMessage.getRequestHeader().getMethod());
+                    sendAndReceive(unauthorisedMessage);
+                    String unauthorisedCachedResponse =
+                            unauthorisedMessage.getResponseBody().toString();
+                    // similarity between authorised and unauthorised response is checked
+                    if (checkSimilarity(unauthorisedCachedResponse, cachedResponse)) {
+                        extensions.add(ext);
+                    }
+                }
+                uri.setPath(path);
+                authorisedMessage.getRequestHeader().setURI(uri);
+                sendAndReceive(authorisedMessage);
+                if (!extensions.isEmpty()) {
+                    String attack = "";
+                    for (String ext : extensions) {
+                        attack = attack + "/test." + ext + ",";
+                    }
+                    newAlert()
+                            .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                            .setMessage(authorisedMessage)
+                            .setOtherInfo(
+                                    Constant.messages.getString(
+                                            "ascanalpha.webCacheDeception.otherinfo"))
+                            .setAttack(attack)
+                            .raise();
+                    return;
+                }
+            } catch (IOException e) {
+                LOG.warn(e.getMessage(), e);
+            }
+        }
+    }
+
+    private boolean checkSimilarity(String a, String b) {
+        LevenshteinDistance distance = new LevenshteinDistance();
+        int levenshteinDistance = distance.apply(a, b);
+        // if response length is less than threshold
+        if (b.length() <= LEVENSHTEIN_THRESHOLD) {
+            if (levenshteinDistance < (int) (b.length() * 0.90)) {
+                return true;
+            }
+        } else {
+            if (levenshteinDistance < LEVENSHTEIN_THRESHOLD) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean initialTest() {
+        HttpMessage authorisedMessage = getNewMsg();
+        try {
+            sendAndReceive(authorisedMessage);
+            URI uri = authorisedMessage.getRequestHeader().getURI();
+            String authorisedResponse = authorisedMessage.getResponseBody().toString();
+
+            HttpMessage unauthorisedMessage = new HttpMessage(uri);
+            unauthorisedMessage
+                    .getRequestHeader()
+                    .setMethod(authorisedMessage.getRequestHeader().getMethod());
+            sendAndReceive(unauthorisedMessage);
+            String unauthorisedResponse = unauthorisedMessage.getResponseBody().toString();
+            // check whether authorised and unauthorised message are similar
+            if (checkSimilarity(unauthorisedResponse, authorisedResponse)) {
```

Comment:
I'd expect this initial test return false in majority of the cases, it's more likely that the original message sent authenticated will be authenticated again (e.g. spidering as user, then scanning as user).

@kingthorin, @psiinon thoughts?

Reply:
Is there a convenient way for scan rules to override or prevent the scanner from authenticating a request?

Reply:
Would using some other library like HttpsURLConnection for unauthorized responses only work here?

Reply:
Yes, but it would bypass other ZAP configurations though.

@kingthorin not reliably on a single message, as workaround the scan rule could remove the user from the `HttpSender` but that could be in use by other scan rules (so some of the requests that are supposed to be authenticated would no longer be). If it's used Forced User mode it requires more tweaks.

Reply:
Understood, thanks for clarifying.

Reply:
I'd not oppose using `HttpsURLConnection` for the time being, though the caveats should be noted in the help.

Reply:
In any case the help should mention that the scan rule expects an user to be used.